### ISSUE TITLE
Clarification on returning array vs object for an individual resource

### ISSUE THREAD

**Issue by chingor13** [json-api/json-api#268]
From #237, a breaking change to how we should return individual resources was accepted.

Previously, we always returned an array for ease of processing. I realize it's from a year ago, but in [#8](https://github.com/json-api/json-api/issues/8#issuecomment-17442480), it seemed like "always return an array" was the accepted thinking.

> BREAKING CHANGE: Singular resource objects SHOULD now be be represented with
> JSON objects instead of arrays. This allows for symmetrical representations in
> request and response documents, as well as PUT/POST requests and PATCH
> operations. It also simplifies implementations which do not support batch
> operations (i.e. they can allow an object and not an array).

This was the explanation and it seemed like there was scattered discussion within #237, but I haven't seen rationale as to why this change was accepted/desired?

**Comment by ahacking** [comment]
I agree, I've been following json-ap with interest but I diverge from it in a number of areas and this is one of them.

I always return an array because I only need to implement one code path. I also only return ids for belongs-to / many-to-one relationships so I never have arrays of links out of a resource so there's never any singular vs many concern there either. There is only ever many-to-one links out of my resources as returning ids on the many side is wasteful and doesn't scale when you have high cardinality relationships, plus queries become much simpler  /foos?bar_id=xxx rather than trying to serialize hundreds or thousands of ids.  This is nothing new, databases maintain the  foreign keys on the belongs to side and  have been doing this for decades because its more efficient.

Maybe some more things to consider...

**Comment by ethanresnick** [comment]
The rationale given in the PR for representing single resources as objects doesn't make sense to me either. (What does using an object have to do with symmetry or supporting `PATCH`?) Meanwhile, I also agree that always responding with an array would make implementation simpler.

But maybe there are other reasons for this single-resources-as-objects representation? Here's one that comes to my mind...For better or worse, the JSON API spec seems designed so that clients can determine whether a relationship is a hasMany or belongsTo, without being explicitly told. (This is why an blank hasMany relationship is represented by an empty array, but a blank belongsTo is represented with null; why the `id` and `ids` keys are used in the object representation of a resource's relationship, as opposed to some common `refs` key, etc.) Maybe this is not something the spec should offer (a schema might be a better fit), but if we accept that it's a legitimate end, then it seems like returning single resources as objects is required to deliver it. For example:

```
{
  "links": {
    "listings.photographs": "http://myapi.com/photographs?listing={listings.id}"
  },
  "listings": {
    "id": "3",
    "title": "Blah"
  }
}
```

Here, the client can only figure out whether listings.photographs is a belongsTo or hasMany based on whether it gets an array or a single document back from `http://myapi.com/photographs?listing=3` request.

**Comment by dgeb** [comment]
> BREAKING CHANGE: Singular resource objects SHOULD now be be represented with
> JSON objects instead of arrays. This allows for symmetrical representations in
> request and response documents, as well as PUT/POST requests and PATCH
> operations. It also simplifies implementations which do not support batch
> operations (i.e. they can allow an object and not an array).

I certainly jammed a lot of reasoning into that one paragraph, which deserves more thorough analysis, especially for a breaking change. I didn't make this change on a whim, but only after a couple months working on a branch that ultimately became #237.

Let me expand upon some of the history and reasoning behind this change.

## Some History

At the time @wycats [commented on #8](https://github.com/json-api/json-api/issues/8#issuecomment-17442480) there was no separation between the primary resource and related resources. Given that compound documents might freely mix primary and related resources, even when a single resource was requested, it seemed like a great simplification to just "always return an array".

However, there were big holes in the approach of mixing primary and related resources: namely, related resources could be of the same type as primary resources. [I commented on this](https://github.com/json-api/json-api/issues/8#issuecomment-17454641) immediately, but we didn't really formalize a solution until introducing the top-level `linked` object.

The introduction of the `linked` object allows the server to clearly distinguish the primary resource from related resources, regardless of type, which allows the client to perform operations such as pagination of a set of primary resources.

## Intent

The introduction of the `linked` object also allows the primary resource in a request AND response to match the intention of the request.

Although it may seem subtle, the following requests are fundamentally different in intent:

```
GET /photos/1
```

```
GET /photos?id=1
```

The first is a request for a single photo with `id === 1`, while the second is a request for a collection of photos that are filtered by `id === 1`.

The first request should either return a single object or 404. The second should return a collection that contains zero or one members.

By returning a single object when a single object has been requested, the response can directly match the intent of the request.

This might be better illustrated with an update request:

```
PUT /photos/1
```

Sending an object instead of an array better matches the intent to update this single resource.

## Symmetry

Returning an object from `GET /photos/1` and expecting an object from `PUT /photos/1` makes for a symmetry between requests and responses.

JSON API's [PATCH support](http://jsonapi.org/format/#patch) was also designed for symmetry with POST / PUT / DELETE. Objects are added to collections, attributes are updated, etc. There is no need to artificially reference a single object as an array in a PATCH operation's `value`.

## Incremental implementations

There are a lot of MAYs in JSON API, which sets a low bar for initial compliance and allows for incremental adoption of more advanced features as needed by an API.

One of these features is the ability to create or update multiple resources in a single request. Since this operation must be transactional, and completely succceed or fail, it may be non-trivial and undesirable to implement.

By distinguishing between objects and arrays, servers can simply reject out of hand any arrays that are sent with a POST or PUT if they choose not to work with transactional changes that span multiple resources. This can lead to cleaner APIs and less confusion on the client and server (I'm sure I'm not the only one who would find an endpoint that requires a single-member array to be a bit unintuitive).

## Lots of related issues

These are just some of the reasons I can think of off the top of my head. There have been others that have arisen in issues, many of which are closed. For instance, should a has-one relationship be updated with an array? Should null ever be returned or should empty arrays be equivalent? I think a lot of these issues stemmed from the unnaturallness of representing everything as an array.

I'm sure that this doesn't appease everyone but hopefully I've explained my perspective on this change in more detail and more clearly.

**Comment by wvteijlingen** [comment]
Thank you for your insight, this does give perspective in why this change was made. I regret however that the spec increased in complexity just for some (imho) small semantic issues. In my opinion, being 'formally' correct here is less important that supplying a simple, consistent spec for which libraries can easily be created.

Client libraries must now include more code paths to deal with these differences. I see _zero_ downsides to always returning an array. It defines a strict standard to which parsers can adhere. No matter what the request was, the response can _always_ be parsed in the same way.

I've been watching JSON API for quite a while now and I've always seen it as a great solution to the api design problem. But, as with any standard, imho the only way this can succeed is through adoption. There must be client and server libraries that facilitate the use of JSON API. I think to stimulate this, the API must be very specific about what is required, and be easy to implement. Adding more complexity makes implementation more complex. I think this leads to less adoption and libraries which diverge from the standard, and thus a failed standard.

**Comment by lgebhardt** [comment]
> Client libraries must now include more code paths to deal with these differences. I see zero downsides to always returning an array. It defines a strict standard to which parsers can adhere. No matter what the request was, the response can always be parsed in the same way.

@wvteijlingen I had the same thoughts about the change, and to some extent I still do. I do see returning both single resources and arrays of resources as making for a more intuitive spec, but I also see the benefits of a simpler response for both generating and parsing. However I think the benefits are minor.

> I think to stimulate this, the API must be very specific about what is required, and be easy to implement. Adding more complexity makes implementation more complex. I think this leads to less adoption and libraries which diverge from the standard, and thus a failed standard.

My initial release of JSONAPI::Resources returned all resources as arrays, based on the earlier version of the spec. I just got through [fixing](https://github.com/cerebris/jsonapi-resources/tree/format) the project to handle arrays and single objects. I expected a bit more trouble, but it was fairly straight forward to come up with the internal rules for when to return what. I think modifying a client library to handle both options should be even easier (but I haven't done it).

**Comment by chingor13** [comment]
@wvteijlingen those were my exact thoughts as well. I was drawn to the initial spec because of the simplicity of knowing that you always get an array back. If you're making a request to an endpoint that you expect to an individual record, you can expect an array of 0 or 1 records.

I haven't yet handled the returning of a single resource in [json_api_client](http://github.com/chingor13/json_api_client) yet, but when I do, I will continue to provide an array interface for consumers of my client.

**Comment by dgeb** [comment]
If the primary resource is always represented by an array, then the following should probably be true for consistency:
- `PUT /photos/1` should accept an array
- `GET /photos/1/links/photographer` should return an array
- `POST /photos/1/links/photographer` should accept an array

Furthermore, it could then be easily argued (and has been) that has-one relationships should always be represented with an array, like so:

```
  {
    "id": "1",
    "title": "Rails is Omakase",
    "links": {
      "author": ["17"]
    }
  }
```

Semantic or no, these implications make me uncomfortable. If you'd like to argue for returning the primary resource as an array, then please make it a comprehensive argument that takes into account the whole scope of the spec by addressing issues like these.

JSON isn't an exceptionally structured format, but I believe we should work with the few structural elements it provides.

**Comment by wvteijlingen** [comment]
Yes, that should then indeed be true :). My view is that resources are always embedded in an array. The spec then specifies that requests to those URLs should only contain 1 resource. It is not entirely semantically correct to allow just 1 object in an array, but it is more consistent and makes (de)serializing easier.

If it is determined that singular resources should be returned as objects without an array, then surely you must also use the correct pluralization throughout the entire spec? Or at least use the singular form for the root key for a singular resource.

``` json
{
  "posts": {
    "id": "1",
    "links": {
       "author": {
          "id": 2,
          "href": "http://example.com/users/2",
          "type": "users"
       }
    }
  },
  "links": {
    "posts.author": "http://example.com/users/{posts.author}"
  }
}
```

should then be

``` json
{
  "post": {
    "id": "1",
    "links": {
       "author": {
          "id": 2,
          "href": "http://example.com/users/2",
          "type": "user"
       }
    }
  },
  "links": {
    "post.author": "http://example.com/users/{post.author}"
  }
}
```

**Comment by ahacking** [comment]
Good discussion, lots of things to consider, well here's my take ..

I've diverged from json-api in my project, I always have a typeless array/bag at the top level called 'data' and each object has its singular type within it. No pluralization issue there.

Endpoints I currently make plural but from a cold computational perspective it would be simpler to just use the same singular string as the type name and avoid the need for inflection rules or additional configuration.

My resources only ever return or accept many-to-one/belongs-to/has-one links to another resource. The one-to-many (ie given a post what are its commments) is resolved by a foreign key search on the related collection (ie /comments?post_id=xx). I avoid the problem that seems to plague most JSON APIs and client libraries of passing lots of ids on the parent as it doesn't scale well when you have thousands of them. Membership/relationships are always managed on the 'belongs to' side.  Because of this my links are always singular and never an array. I also don't segregate links from attributes as it just adds unneeded complication and doesn't allow links out of rich embedded types. Putting links in a separate bag is also at odds with how you want to search, ie doing a /comments?post_id=3 should really be something like /comments?links.post_id=3  

In cases where I return additional resources aka 'side loading', I use a 'linked' bag which is again an array of resources with each resource containing its singular type name.  I don't use typed collection bags as that just adds additional processing overhead on the server which I prefer to avoid. 

The upshot is I have very predicable processing and one way of handling GET of a single resource and GET of a collection and whatever related resources are returned.

**Comment by dgeb** [comment]
Resource `type` is not required to be singular or plural, just consistent throughout the API. The spec uses plural examples because the recommendation is for resource collection names to be used to form resource URLs, and the convention for RESTful APIs is for resource URLs to be plural (e.g. `/posts`). This approach allows the spec to avoid taking a stance on inflection issues.

**Comment by dgeb** [comment]
@ahacking it sounds like your API is very close to compliance with JSON API. [The spec does allow the primary resource to be keyed by the generic key `data`](http://jsonapi.org/format/#document-structure-top-level). Furthermore, there's no requirement that has-many links be expanded in `links`.

My understanding is that you've chosen to go your own way with regards to:
- Excluding the `links` object within resources (#238)
- Always using an array to represent the primary resource (this issue)

I do not think either of these are insurmountable obstacles to compliance, but of course, that's your decision in the end.

On the one hand, I am glad if the spec can provide some guidance for designing an API, even if it is not fully compliant. On the other, I am concerned about the fragmentation of implementations (or rather "near implementations") that erode the value of shared tooling.

**Comment by l8nite** [comment]
@dgeb : Minor nit on the spec re: inflection - the individual resource example uses the singular "author" instead of "authors"

**Comment by dgeb** [comment]
@l8nite `author` is the name of a relationship. Its resource type is `people` throughout the examples in the spec (please let me know if we've overlooked one).

Relationships should be named whatever is most appropriate and don't need to match the resource type. Resource types should be consistent throughout an API. For instance, a `book` may have both an `author` and an `editor`, and the `type` for both should be `people`.

**Comment by ahacking** [comment]
@dgeb. Yes I am close to json-api. I view the embedded issue not one of embedded resources as those embedded things don't have an id/key, they just dont have identity in their own right and are rich/complex data-types which reference other resources. It would be good to resolve #238.

Json -API provides a framework as it doesn't want to be too prescriptive. and like all frameworks there will be many different embodiments, the question is how do we get interoperability given all the choices json-api provides. If i were to develop a client or server today that had to implement every possibility allowed for I dont think I would be happy, but I can implememnt the parts that my api uses. The trouble is thsts what everyone does and why I believe there is room to develop more prescriptive profiles within the choices json-api provides, eg uuids vs server ids, URLs vs ids and link templates, generic vs type collections, singular objects vs arrays and so on. I guess we should ask ourselves does json-api provide a wide enough gamut to support current needs (eg consider all the things that have been points of contention) address the hot spots then develop one or two profiles that are very prescriptive for sake of streamlined implementations and higher probability of interoperability and generic tooling.

**Comment by wvteijlingen** [comment]
> The trouble is thsts what everyone does and why I believe there is room to develop more prescriptive profiles within the choices json-api provides...

You've expressed my feelings exactly @ahacking. I'm writing a client in Swift aiming for full support. While it is a good exercise to get to know Swift better, the amount of possibilities that I have to support makes the library much more complex and leaves more room for errors.

I don't think I need to repeat my sentiments again but I would **really** like to see a JSON API version 2 that does away with a lot of _MAYs_, decreases possibilities and offers a strict spec for interoperability and tooling, while being flexible enough to gain support.

I don't exactly know the best way to open a discussion on this, or whether the core team actually shares this vision. If more people feel this way then maybe we should set some goals, review the current spec and see what can be changed. I would be willing to invest the time and help out.

**Comment by ethanresnick** [comment]
@wvteijlingen and @ahacking have expressed my feelings exactly on the spec as a whole. Hopefully this feedback, which is coming from people who all really want the spec to succeed but have found it too burdensome to implement, will be taken into account. @wvteijlingen Would love your thoughts on #238  too.

On this issue in particular, though, I'm with @dgeb for the reasons he presented [here](https://github.com/json-api/json-api/issues/268#issuecomment-56208576).

**Comment by dgeb** [comment]
As someone working on both [client](https://github.com/orbitjs/orbit.js/blob/master/lib/orbit-common/jsonapi-source.js) and [server](https://github.com/cerebris/jsonapi-resources) implementations of JSON API myself, I understand and share your concerns.

I agree that the proper path to be more prescriptive is via profiles. It's long been my plan to formalize some of the MAYs that exist across JSON API in terms of profile attributes that are defined by the spec.

I'll try to meet with @steveklabnik and @wycats soon to discuss this in detail and will keep everyone in the loop.

**Comment by ethanresnick** [comment]
@dgeb Thanks. Also, I knew you were working on a server-side implementation, but not a client-side one. This simplifies my life so much! How much of the Orbit code is compatible with Ember Data?

**Comment by ahacking** [comment]
@dgeb I know your pain with orbit, really appreciate your effort on this.

As another example I've implemented a high performance ruby json DSL which on my quad core laptop can serialize 1.4  million address objects a second as would be done in separate requests, not aggregate. I'd like to implement a high speed serializer for json-api but I don't want to sacrifice performance trying to support all options or unnecessary features, so pragmatic profiles of the spec is my desirable outcome whilst still allowing json API proper to support a wider range of use cases.

I am thinking rather than gettibg bogged down in choice A vs B, woukd it be better to provide all the options but provide a matrix of all the mays and shoulds. Then the community can rally around profiles they care about and hopefully we will see clustering and wisdom of the crowd occur as tool implementors  implement the profile(s) that best meet their target users.

It would be good to be able to fingerprint an API against a profile rather than some unknown set of shoulds and mays. At that point you could develop a conformance test suite.

**Comment by dgeb** [comment]
@ethanresnick if you're interested in using Orbit and Ember together, check out [Ember-Orbit](https://github.com/orbitjs/ember-orbit), which is an "ember-ified" wrapper over Orbit. Many of the interfaces were inspired by Ember Data, so it should look pretty familiar (although it's not fully compatible because of fundamental differences in architecture).

@ahacking Your ruby lib sounds pretty impressive! Hopefully we can collaborate on a performant server implementation. I also appreciate your suggestions to outline the optional components in JSON API to develop profiles that implementors can rally around. Conformance test suites are a logical next step.

**Comment by ahacking** [comment]
@dgeb. Your json-api resources looks interesting, I think it could be fruitful to join forces.

**Comment by wvteijlingen** [comment]
Awesome to hear that you're open to suggestions/improvements. I've quickly scanned through the spec and threw together a list of things I think can improve. It's not a proposal or anything, just my thoughts on some points:

https://gist.github.com/wvteijlingen/41a9c8aa30e947f76e94

**Comment by dgeb** [comment]
@wvteijlingen glad you're thinking about how to improve the spec :) I scanned your gist and would say that PATCH is important because it allows for multiple operations to be performed transactionally ... but let's not discuss that here. This issue is wandering. Please expand on your thoughts in individual issues (some of those topics already have corresponding issues).

And I'll open a new issue soon to discuss the development of profiles.

Thanks everyone for the good discussion here. I'll leave this issue open for now in case anyone has further input on object vs. array representations.

**Comment by wid** [comment]
Hello All,

I am puzzled by the fuzzyness of the use plurals and singular in the URL paths ( PATCH /article/1/links/author vs PUT /articles/1 ). 

It looks like the spec prefers plural but a plural returning a list is not suitable to define the canonical URI of an object. 

In (Mark Massé, O'REILLY, REST API Design Rulebook ) there is a very clear behavior described in URL Path Design Chapter:
- A singular noun should be used for document name:
  PUT/GET/DELETE /article/1  ==> **Always Returns a single document**
- A plural noun should be used for collection name:
  GET /articles/search?ids=1  ==> **Always Returns a list**
- An special case would be resource creation:
  /articles (POST) ==> Return a 200 + Location header pointing to the newly created resource : 
  Location: http://localhost//article/1

I think this design would solve the array vs object problem.

**Comment by opsb** [comment]
I have to say this is one of the most confusing aspects of the spec for me. I understand that in some cases not aligning with these rules makes it easier to parse payloads but in my opinion it makes the spec far less intuitive and far harder to comprehend(which is probably more of a barrier than actually implementing it).

**Comment by ahacking** [comment]
@wvteijlingen I also commented on your gist but thought I should instead comment here. For 'paging' I leave behind the notion of pages as it is an artificial construct and doesn't work well for infinite scroll use cases. Its better to just have 'count' (ie total count of records) and the server support 'offset' and 'limit' query params and let clients create their own concept of a page. I typically have a 'fetch more' on my client app which can dynamically determine how many records it needs based on average height of elements, screen orientation and velocity. The beauty of offset and limit is that it also supports sparse arrays and you can easily backfill parts that you don't have when those regions come into view, this allows a user to velocity scroll really quickly and you just fetch what is needed for the region they will ultimately land on. Using pages as the fundamental construct really gets in the way vs a more fundamental `offset` and `limit` approach.

**Comment by wvteijlingen** [comment]
Hmm, I never got a notification of your comment. You are right about limit/offset vs pages. I think it would be nice to add such a system to the spec so client libraries can implement paging according to the spec.

**Comment by xiaohanyu** [comment]
> I don't think I need to repeat my sentiments again but I would really like to see a JSON API version 2 that does away with a lot of MAYs, decreases possibilities and offers a strict spec for interoperability and tooling, while being flexible enough to gain support

+1 for this @wvteijlingen 

And I think pagination is a must in API design, so I think it's better for json-api provide some hints about how to express pagination, maybe in top-level `meta` element?

**Comment by dgeb** [comment]
Wow, there's a lot to digest in this thread. I believe that #341(rc2) now addresses most of these concerns, from pagination to a simplification of the spec to reduce the number of MAYs. If there's anything left unaddressed, please open a new more focused issue.
